bpo-30844: selectors: add urgent data to read event

pklanke opened the pull request:
<!-- issue-number: bpo-30844 -->
https://bugs.python.org/issue30844
<!-- /issue-number -->

pklanke committed:
selectors: use predefined events masks in EpollSelector.select

This fixes an inconsistency between base class' select method, which
uses predefined class members _EVENT_READ and _EVENT_WRITE, and
EpollSelector.select, which uses select.EPOLL* events directly.

Signed-off-by: Pim Klanke <pim@protonic.nl>

pklanke committed:
selectors: Add missing exceptional conditions event support

Signed-off-by: Pim Klanke <pim@protonic.nl>

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

vstinner reviewed (commented):
Wow, nice patch! Todo:
* Add a NEWS entry using the blurb tool, see: https://docs.python.org/devguide/committing.html#what-s-new-and-news-entries
* Document the 2 newly added methods in Doc/library/selectors.rst, don't forget ".. versionadded:: 3.7"
* Maybe also document the enhancement in Doc/whatsnew/3.7.rst, in a new "selectors" section

vstinner reviewed (commented):
I would prefer to first focus on the selectors module to have a shorter PR, easier to review. And later extend asyncio.

I mean, keep your asyncio changes somewhere, and revert all asyncio changes here.

pklanke commented:
If I understand correctly, you want me to: 
1) remove all references to asyncio from the original issue (bpo-30844) and remove the changes to asyncio/selector_events.py and related tests from this patch
2) create a second issue regarding asyncio missing the same feature
3) create a second patch solving the second issue

Step 2 and 3 can not exist without this patch. Chapter 11 of the Python Developers Guide, Triaging an issue states under 11.1.10. Dependencies, that dependencies can be listed to be resolved first, before an issue can move forward. Since their is no field marked "dependencies" when creating a new issue, do I just list them in the comment?

vstinner commented:
"create a second patch solving the second issue"

This can wait until this PR is merged ;-)

pklanke commented:
I am sorry, I see now that I have committed a foul (no pun intended). I removed the commit regarding asyncio from my branch and executed a force push to GitHub, something that is clearly listed as not allowed (sorry, newby). Do I need to fix this by reverting the change in the title, adding the commit again and then remove it explicitly and then again change the title, or should I just leave it for now

vstinner commented:
You can use git push --force on your branch, pklanke:fix-issue-30844.

I'm doing that multiple times per day :-) When I'm in my local branch (ex: "fix-issue-30844" in your case), I use "git push haypo HEAD -f", where "haypo" is my Git remote repostory (see "git remote -v").

pitrou commented on Doc/library/selectors.rst:
```diff
@@ -63,6 +63,11 @@ constants below:
    +-----------------------+-----------------------------------------------+
    | :const:`EVENT_WRITE`  | Available for write                           |
    +-----------------------+-----------------------------------------------+
+   | :const:`EVENT_EXCEPT` | Available for urgent data                     |
```
Please call it `EVENT_URGENT` then. `EVENT_EXCEPT` is confusing in Python context (do I get an exception?).

pitrou commented:
As a side note, `kqueue` does not seem to have support for urgent data. Asking for urgent data with `kqueue` should probably fail with a well-defined error (`NotImplementedError`?).

vstinner commented:
"As a side note, kqueue does not seem to have support for urgent data. Asking for urgent data with kqueue should probably fail with a well-defined error (NotImplementedError?)."

Oh. Maybe we should not add add_excepter() on KqueueSelector?

pklanke commented:
"Oh. Maybe we should not add add_excepter() on KqueueSelector?"

You're confusing me; You asked me to remove the asyncio related part from this patch and solely focus on the selectors patch first. add_excepter() is part of the asyncio patch and no longer a part of this patch. On top of that, add_excepter is, was, and will not be added to KqueueSelector.

vstinner commented:
"You're confusing me; You asked me to remove the asyncio related part from this patch and solely focus on the selectors patch first. add_excepter() is part of the asyncio patch and no longer a part of this patch."

Sorry, I didn't read carefully your PR. You're right, I'm wrong :-) Ignore my previous comment. Sadly, it seems like NotImplementedError is the only option for KqueueSelector. It should be carefully documented.

pklanke commented:
The kernel documentations for sysfs states the following:

`If the pin can be configured as interrupt-generating interrupt
and if it has been configured to generate interrupts (see the
description of "edge"), you can poll(2) on that file and
poll(2) will return whenever the interrupt was triggered. If
you use poll(2), set the events POLLPRI and POLLERR. If you
use select(2), set the file descriptor in exceptfds. After
poll(2) returns, either lseek(2) to the beginning of the sysfs
file and read the new value or close the file and re-open it
to read the value.`

In my opinion, sysfs kind of misuses available events to be able to support edge triggered io events, since the read event can not be used (sysfs GPIO fd's are always available for read). However, the patch has no relation to sysfs or TCP, sysfs is only mentioned as a use case for the exceptional conditions event.

Naming the event is not based on what it is used for, but on what lies beneath: man select(2), select_tot(2), poll(2) and epoll(7). IMO 'exceptional conditions event' is the best choice here, since this is the common denominator used by select (which was what we used before the selectors module was added in python 3.4). The term 'urgent data' originates from (e)poll, which some might say could also be used, but this would result in adding add_urgenter() to asyncio's selector_events module (again, not part of this patch anymore)

I really regret using the term 'urgent data' in the documentation and the title now ;-). Can we agree on calling it 'exceptional condition'?

vstinner commented:
Please move the discussion to http://bugs.python.org/issue30844 (we prefer
discusson on bugs.python.org rather than GitHub, and I would prefer to not
discuss at two different places), thanks ;-)

pklanke committed:
selectors: KQueueSelector has no support for urgent data to read

Signed-off-by: Pim Klanke <pim@protonic.nl>

pklanke committed:
selectors: discard fileobj from urgents list on unregister.

Signed-off-by: Pim Klanke <pim@protonic.nl>

1st1 commented on Lib/selectors.py:
```diff
@@ -470,9 +488,11 @@ def select(self, timeout=None):
             for fd, event in fd_event_list:
                 events = 0
-                if event & ~select.EPOLLIN:
+                if event & ~(self._EVENT_READ | self._EVENT_WRITE):
```
Can we precompute the masks to move any unnecessary computation out of the loop?

pklanke replied:
Good point. I'll change this.

pklanke committed:
selectors: precompute the masks used by _PollLikeSelector.select

precomputing the masks moves unnecessary computation out of the loop.

Signed-off-by: Pim Klanke <pim@protonic.nl>

1st1 commented on Lib/selectors.py:
```diff
@@ -429,13 +429,16 @@ def select(self, timeout=None):
             return ready
+        u = ~(self._EVENT_READ | self._EVENT_WRITE)
+        w = ~(self._EVENT_READ | self._EVENT_URGENT)
+        r = ~(self._EVENT_WRITE | self._EVENT_URGENT)
```
Please use more descriptive names, like `r` -> `read_flags`

1st1 commented on Lib/selectors.py:
```diff
@@ -550,7 +597,38 @@ def unregister(self, fileobj):
+            return super().modify(fileobj, events, data)
+
         def select(self, timeout=None):
+            """Perform the actual selection, until some monitored file objects are
```
First docstring line should be a complete sentence < 72 chars long (a lot of tools expect that).

pklanke replied:
You are right, but all I did was copy the original docstring from the abstract base class before editing it. I feel that if I change this line, I should also change the original doc string of the abstract base class.

Then there are a whole bunch of other lines exceeding the maximum line length. I feel that if I change one of them, I should fix them all.

vstinner replied:
Fix them all in that case please ;-)

1st1 commented on Lib/selectors.py:
```diff
@@ -98,7 +99,7 @@ def register(self, fileobj, events, data=None):
         Parameters:
         fileobj -- file object or file descriptor
-        events  -- events to monitor (bitwise mask of EVENT_READ|EVENT_WRITE)
+        events  -- events to monitor (bitwise mask of EVENT_READ|EVENT_WRITE|EVENT_URGENT)
```
This line seems to be longer than 79 chars, right?

vstinner commented on Doc/library/selectors.rst:
```diff
@@ -63,6 +63,11 @@ constants below:
+   +-----------------------+-----------------------------------------------+
+
+   .. versionchanged:: 3.7
+      Added EVENT_URGENT
```
Please add a final "." to your sentence.

vstinner commented on Doc/library/selectors.rst:
```diff
@@ -63,6 +63,11 @@ constants below:
    | :const:`EVENT_WRITE`  | Available for write                           |
    +-----------------------+-----------------------------------------------+
+   | :const:`EVENT_URGENT` | Available for urgent data to read             |
+   +-----------------------+-----------------------------------------------+
```
Hum, since KqueueSelector doesn't support it. It would be nice to list selectors which support it or not. Maybe just mention that KqueueSelector doesn't support urgent events?

vstinner commented:
> @pklanke pklanke closed this 4 minutes ago

Oh, why did you close your PR? Is it a mistake?

terryjreedy commented:
I just ran test suite on freshly updated master on Windows and only failure was test_compileall.  I reran three times by iteself and is passed 3 times.  The Appveyor (windows) test had 15 failures (it is currently rerunning them): see line 5917.  test_selectors is among them, so my guess is that your patch is not correct for Windows, but I cannot help diagnose. I don't know if everything else imports selectors.

pitrou commented on Lib/selectors.py:
```diff
@@ -294,43 +298,52 @@ def __init__(self):
-            r, w, x = select.select(r, w, w, timeout)
-            return r, w + x, []
+        def _select(self, r, w, x, timeout=None):
+            r, _w, _x = select.select(r, w, w|x, timeout)
```
It would be nice to add a comment explaining the `w|x` thing as well as the `_w+_x` below.

pklanke commented:
All checks have passed and no conflicts exist with the base branch. Is there anything more I can or need to to do to get this merged into the master branch?

vstinner commented on Doc/whatsnew/3.7.rst:
```diff
@@ -303,6 +303,11 @@ The flags :const:`re.ASCII`, :const:`re.LOCALE` and :const:`re.UNICODE`
 (Contributed by Serhiy Storchaka in :issue:`31690`.)
 
+selectors
+---------
```
nitpick: please add an empty line.

vstinner commented on Doc/library/selectors.rst:
```diff
@@ -63,6 +63,12 @@ constants below:
+   +-----------------------+-----------------------------------------------+
+
+   .. versionchanged:: 3.7
+      Added EVENT_URGENT.
```
nitpick: use Sphinx formatting:

```
Added :const:`EVENT_URGENT`.
```

pklanke committed:
selectors: Changed confusing EVENT_EXCEPT to EVENT_URGENT

Naming the event for the exceptional conditions event was mainly based on
exceptfds, the 3rd argument of select(). Since except already has a very
distinctive meaning in the world of Python, it is better to replace it by
one of the already accepted synonyms, in this case "urgent data to read"
(used by poll()).

Signed-off-by: Pim Klanke <pim@protonic.nl>

pklanke committed:
selectors: Fix doc strings exceeding the maximum length

Signed-off-by: Pim Klanke <pim@protonic.nl>

pklanke committed:
selectors: Remove evaluation that is always true

Signed-off-by: Pim Klanke <pim@protonic.nl>

MaxwellDupre reviewed (commented):
Got an error when running specific test:
./python -m unittest -v test.test_selectors.test_main
----------------------------------------------------------------------
Ran 114 tests in 18.104s

OK (skipped=40)
Traceback (most recent call last):
  File "/home/me/Documents/cpython/Lib/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/me/Documents/cpython/Lib/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/me/Documents/cpython/Lib/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/home/me/Documents/cpython/Lib/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/home/me/Documents/cpython/Lib/unittest/main.py", line 141, in parseArgs
    self.createTests()
  File "/home/me/Documents/cpython/Lib/unittest/main.py", line 148, in createTests
    self.module)
  File "/home/me/Documents/cpython/Lib/unittest/loader.py", line 219, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/home/me/Documents/cpython/Lib/unittest/loader.py", line 219, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/home/me/Documents/cpython/Lib/unittest/loader.py", line 211, in loadTestsFromName
    (obj, test))
TypeError: calling <function test_main at 0x7ff7a4200338> returned None, not a test

cpython on  fix-issue-30844 [$?] via 🐍 v3.11.0a3+ took 18s
